### PR TITLE
docs: add notes on distributed deployment

### LINKED ISSUE #1979: Document how to handle sessions, load balancing, and scaling in production
Based on community feedback, it would be helpful to add a page under the deploying to production section explaining how to deal with sessions and load balancing. Persistent sessions and a session replication - should you do it?

### PR DESCRIPTION
**Pull request by mcollovati**
Brief notes on scalability, high availability and session replication. Links to Kubernetes Kit as session replication solution.

Closes #1979

### CODE REVIEW

File: articles/production/distributed-deployment.adoc [hunk 1]
```diff
@@ -0,0 +1,66 @@
+---
+title: Distributed Deployment Architecture
+description: Application deployment on multiple servers.
+order: 57
+---
+
+= Distributed Deployment Architecture
+
+Vaadin applications can be deployed on multi-host distributed architectures.
```
Comment (marcushellberg) [inline articles/production/distributed-deployment.adoc:9]: I suggest you add an example here to make it more concrete

File: articles/production/distributed-deployment.adoc [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+This architecture meets the requirements of scalability and high availability, but introduces extra complexity to the deployment processes.
+
+== Scalability
+
+Scalability refers to the ability of the application to handle huge volume of workload and expand in response to an increased traffic it receives.
+Vertical scalability consists in adding more resources to the system so that it meets demand, whereas horizontal scalability means adding more machines hosting different instances of the same application.
+
+Vertical scalability has the advantage of fewer costs and less complicated maintenance.
+However, it is susceptible to downtime, for example during upgrades,
+it introduces a single point of failure, and it should consider the machine's physical upgrade limitations.
```
Comment (marcushellberg) [inline articles/production/distributed-deployment.adoc:19]: Is this still part of the previous sentence, or a new sentence?

File: articles/production/distributed-deployment.adoc [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+== High Availability
+
+High availability is the ability of a system to ensure high performance without interruption of any kind in a given period of time.
+This means that when a server stops working, it should be transparently replaced by a different machine, in a way that the user should be able to continue using the application without any disruption.
+
+For Vaadin application, this scenario requires some kind of session replication, since the session data must be distributed between multiple servers to ensure availability.
+
+Even if the session replication is a requirement for full high availability, the recommendation is to keep using the sticky session approach for as long as a specific server is available.
+This is because the efforts required to make the Vaadin application state stored in the HTTP session fully serializable and distributable to different servers on the network.
```
Comment (marcushellberg) [inline articles/production/distributed-deployment.adoc:43]: Wouldn't it be even more difficult to do this if you haven't taken it into account from the beginning of the project?

File: articles/production/distributed-deployment.adoc [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+
+== Session Replication
+
+To be able to replicate the session on different servers, the Java components used in the UI must implement the [interfacename]`java.io.Serializable` interface.
+Of course, also all class members must be serializable or defined as `transient` fields.
+In addition, the application must be able to handle transient fields during deserialization.
```
Comment (marcushellberg) [inline articles/production/distributed-deployment.adoc:50]: Can you make this more specific, i.e. "the application must be able to initialize transient field values" or something similar? "handle" is a bit vague

File: articles/production/distributed-deployment.adoc [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+The session needs to be serialized and stored in a distributed persistence storage on every request.
+When the user requests are forwarded to a different server, the data must be loaded, deserialized and used to populate the new session.
+
+Implementing an effective session replication for Vaadin application is not easy.
```
Comment (marcushellberg) [inline articles/production/distributed-deployment.adoc:55]: I would change this to a bit less drastic, for instance "Implementing effective session replication for Vaadin applications can be challenging". And then perhaps making the following paragraph 2 bullet points that support the statement.

File: articles/production/distributed-deployment.adoc (hunk 2 above)
Comment (mcollovati) [reply articles/production/distributed-deployment.adoc:19]: A new sentence. The previous one introduces the concepts of vertical and horizontal scalability.
This one is about advantages and disadvantages of vertical scalability

File: articles/production/distributed-deployment.adoc (hunk 1 above)
Comment (mcollovati) [reply articles/production/distributed-deployment.adoc:9]: Added a new sentence, but I'm uncertain if it is the kind of example you have in mind

File: articles/production/distributed-deployment.adoc (hunk 3 above)
Comment (mcollovati) [reply articles/production/distributed-deployment.adoc]: Added a new sentence, to emphasize the additional difficult if the application is not designed to be distributable from the beginning

File: articles/production/distributed-deployment.adoc (hunk 4 above)
Comment (mcollovati) [reply articles/production/distributed-deployment.adoc]: Updated

File: articles/production/distributed-deployment.adoc (hunk 5 above)
Comment (mcollovati) [reply articles/production/distributed-deployment.adoc]: Updated the sentence as suggested and moved it to the beginning.
Previous sentences are now a bullet list, presenting several of the requirement to setup a distributed session

File: articles/production/distributed-deployment.adoc (hunk 2 above)
Comment (marcushellberg) [reply articles/production/distributed-deployment.adoc]: In that case, end the previous sentence with a period and start this one with a capital letter.

File: articles/production/distributed-deployment.adoc (hunk 1 above)
Comment (marcushellberg) [reply articles/production/distributed-deployment.adoc:9]: I was thinking "such as Kubernetes" or something along those lines

**Review by marcushellberg — APPROVED** [review]

File: articles/production/distributed-deployment.adoc (hunk 2 above)
Comment (mcollovati) [reply articles/production/distributed-deployment.adoc]: Ah, sorry. I did not understand that you were talking about the "it introduces ..." part.
It was the same sentence.
But now the section has been changed to use a bullet list

**Pull request merged**